sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+extern crate clap;
+extern crate glib;
+extern crate gstreamer as gst;
+extern crate gstreamer_sdp;
+extern crate gstreamer_sdp_sys;
+extern crate gstreamer_webrtc;
+extern crate rand;
+extern crate ws;
+#[macro_use]
+extern crate serde_json;
+
+use gst::prelude::*;
+use gst::{BinExt, ElementExt};
+use rand::Rng;
+use std::sync::{Arc, Mutex};
+
+#[derive(PartialEq, PartialOrd, Eq, Debug)]
+enum AppState {
+    // AppStateUnknown = 0,
+    AppStateErr = 1,
+    ServerConnecting = 1000,
+    ServerConnectionError,
+    ServerConnected,
+    ServerRegistering = 2000,
+    ServerRegisteringError,
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallError,
+}
+#[derive(Debug)]
+enum Error {
+    GlibError(glib::Error),
+    GlibBoolError(glib::BoolError),
+    GstStateChangeError(gst::StateChangeError),
+}
+
+impl From<glib::Error> for Error {
+    fn from(error: glib::Error) -> Self {
+        Error::GlibError(error)
+    }
+}
+
+impl From<glib::BoolError> for Error {
+    fn from(error: glib::BoolError) -> Self {
+        Error::GlibBoolError(error)
+    }
+}
+
+impl From<gst::StateChangeError> for Error {
+    fn from(error: gst::StateChangeError) -> Self {
+        Error::GstStateChangeError(error)
+    }
+}
+
+const STUN_SERVER: &'static str = "stun-server=stun://stun.l.google.com:19302 ";
+const RTP_CAPS_OPUS: &'static str = "application/x-rtp,media=audio,encoding-name=OPUS,payload=";
+const RTP_CAPS_VP8: &'static str = "application/x-rtp,media=video,encoding-name=VP8,payload=";
```

Comment:
Personally I prefer creating caps explicitly instead of from a string. You could e.g. have functions here that take the remaining missing parameters and then return a `gst::Caps` via `gst::Caps::new_simple()`